errors/errors.go

```diff
@@ -125,32 +126,67 @@ type Error struct {
 }
 
 // Error returns the stored description
-func (e Error) Error() string { return e.desc }
+func (e *Error) Error() string { return e.desc }
 
 // ABCILog returns the stored description, same as Error()
-func (e Error) ABCILog() string { return e.desc }
+func (e *Error) ABCILog() string { return e.desc }
 
 // ABCICode returns the associated ABCICode
-func (e Error) ABCICode() uint32 { return e.code }
+func (e *Error) ABCICode() uint32 { return e.code }
 
 // New returns a new error. Returned instance is having the root cause set to
 // this error. Below two lines are equal
 //   e.New("my description")
 //   Wrap(e, "my description")
 // Allows sprintf format and vararg
-func (e Error) New(description string) error {
+func (e *Error) New(description string) error {
 	return Wrap(e, description)
 }
 
 // Newf is basically New with formatting capabilities
-func (e Error) Newf(description string, args ...interface{}) error {
+func (e *Error) Newf(description string, args ...interface{}) error {
 	return e.New(fmt.Sprintf(description, args...))
 }
 
-// Is is a proxy helper for global Is to be able to easily instantiate and match error codes
-// for example in tests
-func (e Error) Is(err error) bool {
-	return Is(e.New(""), err)
+// Is check if given error instance is of a given kind/type. This involves
+// unwrapping given error using the Cause method if available.
+//
+// Any non weave implementation of an error is tested positive for being
+// ErrInternal.
+func (kind *Error) Is(err error) bool {
+	type causer interface {
+		Cause() error
+	}
+
+	// Reflect usage is necessary to correctly compare with
+	// a nil implementation of an error.
+	if kind == nil {
```

Comment:
Not sure if we need this.
Was kind of happy to not use *Error, which meant these were safely immutible.

Reply:
Using a static method does not make the method call immutable :grin: It means it works on a copy of a structure. You can still modify all the data and if it is a copy of a pointer or a slice then the behavior might be not as expected.

Declaring the `Is` method on a pointer allows to test if an error is `nil`. This makes the tests writing nicer - the `wantErr.Is(gotErr)` test exactly the way you would expect it to.

Does this make sense?

Reply:
With our test cases or prod code we should always know the type on one side of the comparison. It would be ErrUnauthorized.Is(something) and not something.Is(ErrUnauthorized) so that we can safely assume the left side won't be nil. IMHO "value" methods would be better to communicate this than "pointer" methods.
In table tests you can still use pointers in the test specs with nil check in the execution though.